src/java.base/share/classes/java/lang/Long.java

```diff
@@ -234,7 +237,11 @@ public static String toUnsignedString(long i, int radix) {
                 }
                 case 16 -> toHexString(i);
                 case 32 -> toUnsignedString0(i, 5);
-                default -> toUnsignedBigInteger(i).toString(radix);
+                default -> {
+                    long leadingDigits = divideUnsigned(i, radix);  // always positive
+                    int lastDigit = (int)remainderUnsigned(i, radix);
```

Comment:
You can avoid the additional division from `remainderUnsigned()` like so
```suggestion
                    int lastDigit = (int)(i - leadingDigits * radix);
```